### PR TITLE
fix(plugin-chart-table): sort and search time column

### PR DESCRIPTION
**Pull request by ktmud**
🐛 Bug Fix

1. Fix `getTime` undefined error when sorting a time column in table chart. Also refactors how datetime is handled for `react-table` search, sort and render methods.

   

2. Fix spacing issue for table pagination in dashboards
   
   **Before:**

   

   **After:**

   

## Test Plan 

Unit test in superset-ui and E2E test in superset-frontend.

cc @serenajiang @kristw

### CODE REVIEW

File: plugins/plugin-chart-table/src/transformProps.ts [hunk 1]
```diff
@@ -24,14 +24,52 @@ import {
...
+    // `new Date()` is capable of handling null/boolean/undefined, etc;
+    super(value as string);
+
+    this.input = input;
+    this.formatter = formatter;
+  }
+
+  toString(): string {
+    if (this.formatter === String) {
+      return String(this.input);
```
Comment (etr2460) [inline plugins/plugin-chart-table/src/transformProps.ts:68]: maybe this should be `return this.formatter(this.input);`?

File: plugins/plugin-chart-table/src/transformProps.ts [hunk 2]
```diff
@@ -24,14 +24,52 @@ import {
...
+
+  constructor(
+    input: DataRecordValue,
+    { formatter = String, forceUTC = true }: { formatter?: TimeFormatFunction; forceUTC?: boolean },
+  ) {
+    let value = input;
+
+    // force UTC time for timestamps without a timezone
+    if (forceUTC && typeof value === 'string' && REGEXP_TIMESTAMP_NO_TIMEZONE.test(value)) {
+      value = `${value}Z`;
```
Comment (etr2460) [inline plugins/plugin-chart-table/src/transformProps.ts:55]: is this correct? I thought you needed to replace the `T` with a `Z` to convert the string to UTC time

File: plugins/plugin-chart-table/src/transformProps.ts (hunk 1 above)
Comment (ktmud) [reply plugins/plugin-chart-table/src/transformProps.ts:68]: It's basically the same thing?

File: plugins/plugin-chart-table/src/transformProps.ts (hunk 2 above)
Comment (ktmud) [reply plugins/plugin-chart-table/src/transformProps.ts:55]: `T` is the separator between date and time. I don't think I've ever seen timestamps like `2020-07-07Z18:20:02.495`.

File: plugins/plugin-chart-table/src/transformProps.ts [hunk 3]
```diff
@@ -24,14 +24,52 @@ import {
...
 const { PERCENT_3_POINT } = NumberFormats;
 const TIME_COLUMN = '__timestamp';
-const toString = (x: DataRecordValue) => String(x);
+const REGEXP_TIMESTAMP_NO_TIMEZONE = /T(\d{2}:){2}\d{2}$/;
+
+/**
+ * Extended Date object with custom formatter.
+ * By default `String(date)` returns the raw input.
+ */
+export class DateWithFormatter extends Date {
```
Comment (kristw) [inline plugins/plugin-chart-table/src/transformProps.ts:42]: could this be separate to its own file?

**Comment by ktmud** [comment]
> I am not sure about extending `Date` but seems like ok for the requirement.

Yeah, I wanted to use `Date.getTime()` value for sorting, but the formatted string for searching. This is the cleanest solution I can think of.

**Pull request merged**